endpoints/openrtb2/amp_auction.go

```diff
@@ -578,21 +588,46 @@ func setAmpExt(site *openrtb2.Site, value string) {
 	}
 }
 
-func readPolicy(consent string) (privacy.PolicyWriter, error) {
-	if len(consent) == 0 {
+// setRegExtGDPR sets regs.ext.gdpr to either 0 or 1 only if it was set in the request query
+func setRegExtGDPR(gdprApplies *bool, req *openrtb2.BidRequest) error {
+	if gdprApplies == nil {
+		return nil
+	}
+
+	gdpr := int8(0)
+	if *gdprApplies {
+		gdpr++
+	}
+
+	reqWrap := &openrtb_ext.RequestWrapper{BidRequest: req}
+	if regsExt, err := reqWrap.GetRegExt(); err == nil {
+		regsExt.SetGDPR(&gdpr)
+	} else {
+		return err
+	}
+	return reqWrap.RebuildRequest()
+}
+
+func (pr policyReader) readPolicy(ampParams amp.Params, req *openrtb2.BidRequest) (privacy.PolicyWriter, error) {
+
+	if err := setRegExtGDPR(ampParams.GdprApplies, req); err != nil {
+		return privacy.NilPolicyWriter{}, err
+	}
+
+	if len(ampParams.Consent) == 0 {
 		return privacy.NilPolicyWriter{}, nil
 	}
 
-	if gdpr.ValidateConsent(consent) {
-		return gdpr.ConsentWriter{consent}, nil
+	if pr.gdpr.ValidateConsent(ampParams.Consent) || ampParams.ConsentType == amp.TCF2 {
```

Comment:
Please update this logic to:

```
if ampParams.ConsentType == none/unknown/unspecified {
  ... code as it always was with existing error and with the gdpr consent writer now also writing the gdpr flag ...
} else {
  switch ampParams.ConsentType {
     case amp.TCF1: ... ignore or error, what does PBS-Java do? ....
     case amp.TCF2: ... validate consent, if error, return error ....
     case amp.USPrivacy: ... validate consent, if error, return error ....
     default: .. error or ignore, unknown consent string type ...
  }
}
```

If the consent type is specified, it should be strictly followed such that TCF1 is ignored or results in an error, TCF2

Reply:
Yep, this part was totally wrong. Logic has been corrected

Reply:
Looks like you're missing the first if condition, to fallback to previous logic when the consent type is not provided.